Rework qg data loading and processing design

mxschmdt opened the pull request:
This change separates data loading from processing for qg so that any dataset in the appropriate format can be used for generating questions and to allow for simple extension to other modalities.

- There are now separate processors depending on the modality.
- Also, the tests have been updated accordingly.
- I've updated the run_qg.py script to reflect the changes.

I've also fixed a few minor issues, like exceptions not being raised.

Let me know if something should be changed.

mxschmdt commented:
The tests seem to fail due to tableqa test errors - I will check again once they have been fixed.

bhavani105 commented:
> The tests seem to fail due to tableqa test errors - I will check again once they have been fixe

jdpsen reviewed (changes requested):
Having a modality param in QGDataLoader breaks the current passage_qg_training notebook. Please rewrite it w.r.t to these changes. 

(2) tydiqa dataloader is different from squad dataloader, because for tydiqa we were using the additional param of "secondary task" while loading this data. W/o the tydi specific processor, it now becomes a user's responsibility to remember/provide that while calling QGProcessor. It is better to check for dataset name and if its tydi include the additional param of "secondary task" from QGProcessor.

jdpsen reviewed (commented):
Why do we need to remove QGTrainer class ? I prefer the design choice of having a common QGTrainer class in PrimeQA which builds on top of HF trainer class. The idea is that any changes related to QG training can be done in the common QGTrainer class once and the same changes would automatically propagate to all other modules using QGTrainer. If we directly built on HF trainer and we need some changes in future, it would mean all the modules using HF Trainer would individually need to incorporate the same changes in multiple places

mxschmdt commented:
> (1) Having a modality param in QGDataLoader breaks the current passage_qg_training notebook. Please rewrite it w.r.t to these changes.
> 
> (2) tydiqa dataloader is different from squad dataloader, because for tydiqa we were using the additional param of "secondary task" while loading this data. W/o the tydi specific processor, it now becomes a user's responsibility to remember/provide that while calling QGProcessor. I suggest we check for dataset name in QGProcessor too and if its tydi include the additional param of "secondary task" from QGProcessor itself w/o needing user to provide it.

I wasn't aware of the notebooks so I updated them now.
I also added your requested TyDi QA change, and extended the QGProcessor class to be able to receive a config and split on initialisation.

> Why do we need to remove QGTrainer class ? I prefer the design choice of having a common QGTrainer class in PrimeQA which builds on top of HF trainer class. The idea is that any changes related to QG training can be done in the common QGTrainer class once and the same changes would automatically propagate to all other modules using QGTrainer. If we directly built on HF trainer and we need some changes in future, it would mean all the modules using HF Trainer would individually need to incorporate the same changes in multiple places

I was replacing it with a trainer for other changes I did not push so far since the QGTrainer did not add any functionality.
I can re-add it again if you want?

jdpsen commented:
> > (1) Having a modality param in QGDataLoader breaks the current passage_qg_training notebook. Please rewrite it w.r.t to these changes.
> > (2) tydiqa dataloader is different from squad dataloader, because for tydiqa we were using the additional param of "secondary task" while loading this data. W/o the tydi specific processor, it now becomes a user's responsibility to remember/provide that while calling QGProcessor. I suggest we check for dataset name in QGProcessor too and if its tydi include the additional param of "secondary task" from QGProcessor itself w/o needing user to provide it.
> 
> I wasn't aware of the notebooks so I updated them now. I also added your requested TyDi QA change, and extended the QGProcessor class to be able to receive a config and split on initialisation.
> 
> > Why do we need to remove QGTrainer class ? I prefer the design choice of having a common QGTrainer class in PrimeQA which builds on top of HF trainer class. The idea is that any changes related to QG training can be done in the common QGTrainer class once and the same changes would automatically propagate to all other modules using QGTrainer. If we directly built on HF trainer and we need some changes in future, it would mean all the modules using HF Trainer would individually need to incorporate the same changes in multiple places
> 
> I was replacing it with a trainer for other changes I did not push so far since the QGTrainer did not add any functionality. I can re-add it again if you want?
Yes, lets re-add the QGTrainer. About the additional functionality you will need, can we put it in the QGTrainer itself (of course keeping it backward compatible), so that they can be reused by other future modules as well ?

mxschmdt commented:
> > > (1) Having a modality param in QGDataLoader breaks the current passage_qg_training notebook. Please rewrite it w.r.t to these changes.
> > > (2) tydiqa dataloader is different from squad dataloader, because for tydiqa we were using the additional param of "secondary task" while loading this data. W/o the tydi specific processor, it now becomes a user's responsibility to remember/provide that while calling QGProcessor. I suggest we check for dataset name in QGProcessor too and if its tydi include the additional param of "secondary task" from QGProcessor itself w/o needing user to provide it.
> > 
> > 
> > I wasn't aware of the notebooks so I updated them now. I also added your requested TyDi QA change, and extended the QGProcessor class to be able to receive a config and split on initialisation.
> > > Why do we need to remove QGTrainer class ? I prefer the design choice of having a common QGTrainer class in PrimeQA which builds on top of HF trainer class. The idea is that any changes related to QG training can be done in the common QGTrainer class once and the same changes would automatically propagate to all other modules using QGTrainer. If we directly built on HF trainer and we need some changes in future, it would mean all the modules using HF Trainer would individually need to incorporate the same changes in multiple places
> > 
> > 
> > I was replacing it with a trainer for other changes I did not push so far since the QGTrainer did not add any functionality. I can re-add it again if you want?

> Yes, lets re-add the QGTrainer. About the additional functionality you will need, can we put it in the QGTrainer itself (of course keeping it backward compatible), so that they can be reused by other future modules as well ?

I've re-added it.
The trainer I will add is basically also inheriting from Seq2SeqTrainer.

vishwajeetkumar93 commented on primeqa/qg/processors/data_loader.py:
```diff
@@ -1,46 +1,61 @@
+from datasets import Dataset, DatasetDict, load_dataset
+from primeqa.qg.processors.table_qg.sql_processor import SqlProcessor
+
+from .passage_qg.qg_processor import QGProcessor
```
A standard naming convention and module structure  could be used for processors like prime.qg.passage_qg import QGPreprocessor

mxschmdt replied:
I've committed the requested changes.

vishwajeetkumar93 commented on primeqa/qg/trainers/qg_trainer.py:
```diff
@@ -1,18 +1,5 @@
-                eval_dataset=valid_dataset,
-                data_collator=data_collator,
-                **kwargs)
+    """ The trainer class for QG. All related functionality should go to this class. """
```
Are we using a trainer or not?

mxschmdt replied:
@jdpsen was asking to re-add it so it is there again in case someone will add some functionality for QG related training or inference.

jdpsen replied:
Don't we want to initialize our trainer here in case we have the option of introducing new params in future.
@vishwajeetkumar93 - any thoughts ?
@mxschmdt : currently it works fine, but I would still prefer a QGTrainer with its own init() here to be future ready. Will that conflict with you current changes ?

mxschmdt replied:
> @mxschmdt : currently it works fine, but I would still prefer a QGTrainer with its own init() here to be future ready. Will that conflict with you current changes ?

No, that would work since I add a separate trainer, maybe we can also merge them (but the current QG implementation does the prediction/generation in the class QGmodel).
If you want to add new functionality you can always add the `init` method, or would you like to add a no-op `init` right now?

jdpsen reviewed (commented):
When I'm trying to run the passageqg_training notebook, I'm hitting a LookupError in evaluate func. of transformers/trainer_seq2seq.py . Che
<img width="1270" alt="Screenshot 2022-09-03 at 12 18 53 AM" src="https://user-images.githubusercontent.com/10701019/188218605-5f8a13f6-d5a0-4899-a3eb-8a1168cbccf3.png">
ck the screenshot for more clue 

Guessing the changes are breaking some existing functionality. Please recheck.

mxschmdt commented:
> When I'm trying to run the passageqg_training notebook, I'm hitting a LookupError in evaluate func. of transformers/trainer_seq2seq.py . Che <img alt="Screenshot 2022-09-03 at 12 18 53 AM" width="1270" src="https://user-images.githubusercontent.com/10701019/188218605-5f8a13f6-d5a0-4899-a3eb-8a1168cbccf3.png"> ck the screenshot for more clue
> 
> Guessing the changes are breaking some existing functionality. Please recheck.

I cannot reproduce the issue, is there more information as to what the LookupError refers to? Is it `self.args.generation_max_length`?
Anyway, it is in the transformers package, maybe try to re-install the pinned transformers version?

jdpsen commented:
@mxschmdt : The previous error has been handled now. Your changes works now without any issues. 
I still think having a QGTrainer with its own init is the best option going forward, left a comment about that. If it doesn't conflict with your changes, can we do it in this PR itself ? We should e able to merge after that.

mxschmdt commented:
> @mxschmdt : The previous error has been handled now. Your changes works now without any issues. I still think having a QGTrainer with its own init is the best option going forward, left a comment about that. If it doesn't conflict with your changes, can we do it in this PR itself ? We should be able to merge after that.

Alright, I just updated the PR.

jdpsen reviewed (approved):
LGTM

bhavani105 reviewed (approved):
LGTM
Based on @jdpsen @vishwajeetkumar93 comments
